rpy2_arrow/pyarrow_rarrow.py

```diff
@@ -191,28 +191,37 @@ def rarrow_to_py_field(
     return pyarrow.lib.Field._import_from_c(_pyarrow_ptr(schema_ptr))
 
 
-def pyarrow_to_r_table(
-        obj: 'pyarrow.lib.Table',
-        py2rpy: typing.Optional[
-            conversion.Converter] = None
-):
+_as_arrow_table_from_stream_ptr = rinterface.evalr(
+    """
+    function(ptr) {
+      arrow::as_arrow_table(arrow::RecordBatchReader$import_from_c(ptr))
+    }
+    """
+)
+
+
+def _pyarrow_table_to_r_table_ri(tbl: pyarrow.lib.Table) -> rinterface.SexpEnvironment:
+    """Create an R `arrow` Table to mirror an Arrow Table in `pyarrow`."""
+    reader_ptr = ffi.new('struct ArrowArrayStream*')
+    tbl.to_reader()._export_to_c(_pyarrow_ptr(reader_ptr))
+    return _as_arrow_table_from_stream_ptr(str(_pyarrow_ptr(reader_ptr)))
+
+
+def pyarrow_table_to_r_table(obj: pyarrow.lib.Table):
     """Create an R `arrow::Table` object from a pyarrow Table.
 
     This is sharing the C/C++ object between the two languages.
     The returned object depends on the active conversion rule in
     rpy2. By default it will be an `rpy2.robjects.Environment`.
-    """
 
-    if py2rpy is None:
-        py2rpy = converter
-    # TODO: this is using the converter defined in the module,
-    # not the converter currently in rpy2.robjects.conversion.
-    kwargs = dict(
-        (k, converter.py2rpy(v))
-        for k, v in zip(obj.schema.names, obj.columns)
-    )
-    kwargs['schema'] = pyarrow_to_r_schema(obj.schema)
-    return rarrow.Table['create'](**kwargs)
+    :param:obj: A :class:`pyarrow.lib.Table`.
+    :return: The result of conversion rules for R `arrow` `Table` objects.
+    By default a :class:`rpy2.robjects.Environment`.
+    """
+    # `res` is a low-level (rinterface-level) rpy2 object. This is an
+    # rpy2 robject-level function. Use the conversion.
+    res = _pyarrow_table_to_r_table_ri(obj)
+    return conversion.get_conversion().py2rpy(res)
```

Comment:
Asking to learn: Not 100% sure why this is necessary. This is already a pointer to an object in the R session. Does this just expose that pointer in a more structured way to Python?

Reply:
`conversion.get_conversion()` is returning the currently active object converter. It is then calling its method `py2rpy`. By default (that is when using the converter coming with `rpy2.robjects`) this is turn a `res` that it of type `rpy2.rinterface.SexpEnvironment` into an `rpy2.robjects.Environment`. Behind the scene the same underlying R objects / pointer to R object is used to instantiate the Python-rpy2 proxy class for it.

However, should one have a conversion rule for R R6 objects added to a converter and that converter be the active one at the same this line is evaluated it will turn the `rpy2.rinterface.SexpEnvironment` into whatever that converter decides it should be.

Did I understand the question?

Reply:
Yes, thank you. Basically this is just to enable any R6 class hooks, such as those you defined in another file, to be activated. Makes sense!